File: python_version/read_voyager.py
import numpy as np

class AtmosphereReader:
    STEFBOLTZ = 5.67051e-5  # erg/cm^2/s/K^4

    def __init__(self, mw_atmos=2.2):
        self.r_atmos = 8.3143e7 / mw_atmos

    def read_voyager(self, teff, grav, nzp1, t_top, p_top):
        """
        Calculate atmospheric profiles for Galileo probe sounding.
        Arguments:
            teff: effective temperature (K)
            grav: gravitational acceleration (cm/s^2)
            nzp1: number of layer edges
            t_top: array of temperatures at layer edges (K), length nzp1
            p_top: array of pressures at layer edges (bars), length nzp1
        Returns:
            nz: number of layers
            z: altitudes at layer midpoints (cm), length nz
            z_top: altitudes at layer edges (cm), length nzp1
            p: pressures at layer midpoints (dyne/cm^2), length nz
            t: temperatures at layer midpoints (K), length nz
            chf: convective heat flux (erg/cm^2/s), length nz
        """
        # Convert p_top from bars to dyne/cm^2
        p_top = np.array(p_top) * 1e6
        t_top = np.array(t_top)
        nz = nzp1 - 1

        z = np.zeros(nz)
        z_top = np.zeros(nzp1)
        p = np.zeros(nz)
        t = np.zeros(nz)
        chf = np.zeros(nz)

        z_top[nz] = 0.0

        for iz in range(nz - 1, -1, -1):
            itop = iz
            ibot = iz + 1
            dlnp = np.log(p_top[ibot] / p_top[itop])
            p[iz] = 0.5 * (p_top[itop] + p_top[ibot])
            scale_h = self.r_atmos * t_top[ibot] / grav
            dz_layer = scale_h * dlnp
            z_top[iz] = z_top[ibot] + dz_layer
            dtdz = (t_top[itop] - t_top[ibot]) / dz_layer
            dz_pmid = scale_h * np.log(p_top[ibot] / p[iz])
            z[iz] = z_top[ibot] + dz_pmid
            t[iz] = t_top[ibot] + dtdz * dz_pmid

        for iz in range(nz):
            chf[iz] = self.STEFBOLTZ * teff ** 4

        return nz, z, z_top, p, t, chf

# Example usage:
# reader = AtmosphereReader(mw_atmos=2.2)
# nz, z, z_top, p, t, chf = reader.read_voyager(teff, grav, nzp1, t_top, p_top)

File: model/read_voyager.f
      subroutine read_voyager( input_file, 
     $   grav, teff, nz, z, z_top, p, p_top, t, t_top, chf )
c
c   Read (and define) atmospheric properties for input format of
c   Galileo probe sounding (variables defined in calling program)
c
c   A. Ackerman Aug-2000
c
      implicit none


c   Include common data shared with eddysed()

      include 'globals.h'


c   Declare local storage

      integer nz, iz, nskip, idummy, nzp1, nmod, itop, ibot, k
      double precision z(MAXNZ)
      double precision p(MAXNZ)
      double precision t(MAXNZ)
      double precision z_top(MAXNZ+1)
      double precision p_top(MAXNZ+1)
      double precision t_top(MAXNZ+1)
      double precision chf(MAXNZ)
      double precision grav, teff, r_atmos, t_in, p_in, dlnp
      double precision scale_h, dz_layer, dz_pmid, dtdz
      character*(*) input_file
      character*1 dumstr


c   Open input file for reading atmospheric profiles

      open( LUNIO, file=input_file, status='old' )


c   Read effective temperature (teff, in K),
c   gravitational acceleration (grav, cm/s^2),
c   and number of layer edges (nzp1)

      read(LUNIO,*) teff
      read(LUNIO,*) grav
      read(LUNIO,*) nzp1

c   Skip text column headings

      nskip = 1
      do iz = 1, nskip
        read(LUNIO,*) dumstr
      enddo


c   Read the data

      if( nzp1-1 .gt. MAXNZ )then 
        print*, 'read_voyager(): nz > MAXNZ'
        stop 1
      endif

c   number of layers to average over
      nmod = 1

      read(LUNIO,*) t_in, p_in
      t_top(1) = t_in
      p_top(1) = p_in*1e6           ! convert bars to dyne/cm^2

      do iz = 1, nzp1-1
        read(LUNIO,*) t_in, p_in
        if( mod(iz,nmod) .eq. 0 )then 
          k = iz / nmod + 1
          t_top(k) = t_in
          p_top(k) = p_in*1e6
        endif
      enddo
      nz = (nzp1 - 1) / nmod

      close(LUNIO)


c   Get pressure thickness and P, T at pressure mid-pt of layer
c   and hydrostatically calculate altitudes

c   should pass atmospheric molecular weight through argument list
      r_atmos = 8.3143e7 / 2.2

      z_top(nz+1) = 0.

      do iz = nz, 1, -1
        itop = iz 
        ibot = iz + 1
        dlnp = log( p_top(ibot)/p_top(itop) )
        p(iz) = 0.5*( p_top(itop) + p_top(ibot) )
        scale_h = r_atmos * t_top(ibot) / grav
        dz_layer = scale_h * dlnp
        z_top(iz) = z_top(ibot) + dz_layer
        dtdz = ( t_top(itop) - t_top(ibot) ) / dz_layer
        dz_pmid = scale_h * log( p_top(ibot)/p(iz) )
        z(iz) = z_top(ibot) + dz_pmid
        t(iz) = t_top(ibot) + dtdz*dz_pmid
      enddo


c   convective heat flux 

      do iz = 1, nz
        chf(iz) = STEFBOLTZ * teff**4
      enddo


      return
      end
